### ISSUE TITLE
BUG: invalid value encountered in divide

### ISSUE THREAD

**Issue by CloseChoice** [shap/shap#3291]
### Issue Description

When using categorical features which are which are more than 50 examples have the same value count (which have the same shap values), the dependence_plot runs into the warning "invalid value encountered in divide". This happens because one we are [sorting](https://github.com/shap/shap/blob/85468af8de52d2c667c2793cd308c9a2aaad637c/shap/utils/_general.py#L122-L125) the shap values by the feature values and then we just use 50 examples and calculate the [cross correlation between them](https://github.com/shap/shap/blob/85468af8de52d2c667c2793cd308c9a2aaad637c/shap/utils/_general.py#L126-L137). 

A solution to this is, that we could use more than 50 examples (making the execution slower) or remove the sorting (is there any reason for that?

This might be related to https://github.com/shap/shap/issues/2647

### Minimal Reproducible Example

```python
import matplotlib.pylab as pl
import numpy as np
import xgboost
from sklearn.model_selection import train_test_split

import shap

X,y = shap.datasets.adult()
X_display,y_display = shap.datasets.adult(display=True)

# create a train/test split
X = X.iloc[:500, :]
y = y[:500]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=7)
d_train = xgboost.DMatrix(X_train, label=y_train)
d_test = xgboost.DMatrix(X_test, label=y_test)

# train final model on the full data set
params = {
    "eta": 0.05,
    "max_depth": 1,
    "objective": "binary:logistic",
    "subsample": 0.5,
    "base_score": np.mean(y_train),
    "eval_metric": "logloss"
}
model_ind = xgboost.train(params, d_train, 5000, evals = [(d_test, "test")], verbose_eval=100, early_stopping_rounds=20)

shap_values_ind = shap.TreeExplainer(model_ind).shap_values(X)

name = X_train.columns[0]  # 'Age'
shap.dependence_plot(name, shap_values_ind, X, display_features=X_display)
```

### Traceback

```shell
invalid value encountered in divide
invalid value encountered in divide
```

### Expected Behavior

This warning should not be thrown

### Bug report checklist

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [X] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [X] I'd be interested in making a PR to fix this bug

### Installed Versions

<details>

absl-py==2.0.0
alabaster==0.7.13
astor==0.8.1
asttokens==2.4.0
astunparse==1.6.3
attrs==23.1.0
autograd==1.6.2
autograd-gamma==0.5.0
Babel==2.12.1
backcall==0.2.0
beautifulsoup4==4.12.2
bleach==6.0.0
cachetools==5.3.1
catboost==1.2.2
certifi==2023.7.22
cfgv==3.4.0
charset-normalizer==3.2.0
cloudpickle==2.2.1
cmake==3.27.5
comm==0.1.4
contourpy==1.1.1
coverage==7.3.1
cycler==0.11.0
debugpy==1.8.0
decorator==5.1.1
defusedxml==0.7.1
distlib==0.3.7
docutils==0.18.1
exceptiongroup==1.1.3
executing==1.2.0
fancycompleter==0.9.1
fastjsonschema==2.18.0
filelock==3.12.4
flatbuffers==23.5.26
fonttools==4.42.1
formulaic==0.6.4
fsspec==2023.9.2
future==0.18.3
gast==0.4.0
google-auth==2.23.0
google-auth-oauthlib==1.0.0
google-pasta==0.2.0
gpboost==1.2.3
graphviz==0.20.1
grpcio==1.58.0
h5py==3.9.0
huggingface-hub==0.17.2
identify==2.5.29
idna==3.4
imageio==2.31.3
imagesize==1.4.1
iniconfig==2.0.0
interface-meta==1.3.0
ipykernel==6.25.2
ipython==8.15.0
ipywidgets==8.1.1
jedi==0.19.0
Jinja2==3.1.2
joblib==1.3.2
jsonschema==4.19.1
jsonschema-specifications==2023.7.1
jupyter_client==8.3.1
jupyter_core==5.3.1
jupyterlab-pygments==0.2.2
jupyterlab-widgets==3.0.9
keras==2.13.1
kiwisolver==1.4.5
lazy_loader==0.3
libclang==16.0.6
lifelines==0.27.8
lightgbm==4.1.0
lime==0.2.0.1
lit==16.0.6
llvmlite==0.41.0
Markdown==3.4.4
MarkupSafe==2.1.3
matplotlib==3.8.0
matplotlib-inline==0.1.6
mistune==3.0.1
mpmath==1.3.0
nbclient==0.8.0
nbconvert==7.8.0
nbformat==5.9.2
nbsphinx==0.9.2
nest-asyncio==1.5.8
networkx==3.1
ngboost==0.4.1
nodeenv==1.8.0
numba==0.58.0
numpy==1.24.3
numpydoc==1.5.0
nvidia-cublas-cu11==11.10.3.66
nvidia-cuda-cupti-cu11==11.7.101
nvidia-cuda-nvrtc-cu11==11.7.99
nvidia-cuda-runtime-cu11==11.7.99
nvidia-cudnn-cu11==8.5.0.96
nvidia-cufft-cu11==10.9.0.58
nvidia-curand-cu11==10.2.10.91
nvidia-cusolver-cu11==11.4.0.1
nvidia-cusparse-cu11==11.7.4.91
nvidia-nccl-cu11==2.14.3
nvidia-nvtx-cu11==11.7.91
oauthlib==3.2.2
opencv-python==4.8.0.76
opt-einsum==3.3.0
packaging==23.1
pandas==1.4.1
pandocfilters==1.5.0
parso==0.8.3
pdbpp==0.10.3
pexpect==4.8.0
pickleshare==0.7.5
Pillow==10.0.1
platformdirs==3.10.0
plotly==5.17.0
pluggy==1.3.0
pre-commit==3.4.0
prompt-toolkit==3.0.39
protobuf==3.20.3
psutil==5.9.5
ptyprocess==0.7.0
pure-eval==0.2.2
py4j==0.10.9.7
pyasn1==0.5.0
pyasn1-modules==0.3.0
Pygments==2.16.1
pyod==1.1.0
pyparsing==3.1.1
pyrepl==0.9.0
pyspark==3.4.1
pytest==7.4.2
pytest-cov==4.1.0
pytest-mpl==0.16.1
python-dateutil==2.8.2
pytz==2023.3.post1
PyWavelets==1.4.1
PyYAML==6.0.1
pyzmq==25.1.1
referencing==0.30.2
regex==2023.8.8
requests==2.31.0
requests-oauthlib==1.3.1
rpds-py==0.10.3
rsa==4.9
ruff==0.0.291
safetensors==0.3.3
scikit-image==0.21.0
scikit-learn==1.3.1
scipy==1.11.2
sentencepiece==0.1.99
-e git+ssh://git@github.com/CloseChoice/shap.git@722b8826912e6dab28ae5bfb5128865cb6339fa0#egg=shap
six==1.16.0
slicer==0.0.7
snowballstemmer==2.2.0
soupsieve==2.5
Sphinx==6.2.1
sphinx-github-changelog==1.2.1
sphinx-rtd-theme==1.2.2
sphinxcontrib-applehelp==1.0.7
sphinxcontrib-devhelp==1.0.5
sphinxcontrib-htmlhelp==2.0.4
sphinxcontrib-jquery==4.1
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.6
sphinxcontrib-serializinghtml==1.1.9
stack-data==0.6.2
sympy==1.12
tenacity==8.2.3
tensorboard==2.13.0
tensorboard-data-server==0.7.1
tensorflow==2.13.0
tensorflow-estimator==2.13.0
tensorflow-io-gcs-filesystem==0.34.0
termcolor==2.3.0
threadpoolctl==3.2.0
tifffile==2023.9.18
tinycss2==1.2.1
tokenizers==0.13.3
tomli==2.0.1
torch==2.0.1
torchvision==0.15.2
tornado==6.3.3
tqdm==4.66.1
traitlets==5.10.0
transformers==4.33.2
triton==2.0.0
typing_extensions==4.5.0
tzdata==2023.3
urllib3==1.26.16
virtualenv==20.24.5
wcwidth==0.2.6
webencodings==0.5.1
Werkzeug==2.3.7
widgetsnbextension==4.0.9
wmctrl==0.5
wrapt==1.15.0
xgboost==1.7.6

</details>
